Direct connect for `OnionMessage` sending

jkczyz opened the pull request:
Not all nodes support onion messages, so it may not be possible for a `MessageRouter` to find a complete path to an `OnionMessage`'s destination. A direct connection to the first node in an `OnionMessagePath` is therefore required. This PR adds a `ConnectionNeeded` event to signal to users to connect to a node.

When sending an `OnionMessage` or replying to one -- but not when forwarding -- `OnionMessenger` will buffer messages for immediate recipients that are not peers and upon connection make them available for processing as normal. A new `release_pending_connections` method is added to the `OnionMessageHandler` trait for the `PeerManager` to call and provide the necessary `ConnectionNeeded` events. Thus, `PeerManager` now implements `EventsProvider` and is called in `BackgroundProcessor` accordingly.

`PeerManager` calls the new method in its own `timer_ticks_occurred` method. This allows `OnionMessenger` to drop any buffered messages for nodes that haven't been connected in a certain number of timer ticks.

TODO: Needs some test coverage.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -148,13 +148,67 @@ where
 }
 
+/// [`OnionMessage`]s buffered to be sent.
+enum OnionMessageBuffer {
```
Was thinking `OnionMessageRecipient` may be a better name.

jkczyz replied:
Renamed as mentioned in a separate commit.

jkczyz commented on lightning/src/events/mod.rs:
```diff
@@ -1173,6 +1190,12 @@ impl Writeable for Event {
 					(0, payment_id, required),
 				})
 			},
+			&Event::ConnectionNeeded { ref node_announcement } => {
+				35u8.write(writer)?;
+				write_tlv_fields!(writer, {
+					(0, node_announcement, required),
+				})
+			},
```
I guess we shouldn't need to serialize these since we don't save buffered messages across restarts.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -516,18 +583,23 @@ where
 	pub fn send_onion_message<T: OnionMessageContents>(
```
Wondering if this should still be `pub`? We send through the handlers, so this is not used internally outside of this module. In this PR, we need to examine the return value to decide if a connection is needed. So users calling this won't trigger a connection request nor will the message be buffered for themexcept if we previously decided that a connection is needed.

jkczyz replied:
The interface here isn't congruent with other sending, FWIW. This takes an `OnionMessagePath` instead of a `Destination`, so the assumption is that the `MessageRouter` has already been called prior to calling `send_onion_message`. We'd probably want to do message routing internally, I'd imagine?

jkczyz commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -2396,6 +2400,13 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, OM: Deref, L: Deref, CM
 			}
 		}
+
+		let nodes_to_connect =
+			self.message_handler.onion_message_handler.release_pending_connections();
+		let mut pending_events = self.pending_events.lock().unwrap();
+		for node_announcement in nodes_to_connect {
+			pending_events.push(Event::ConnectionNeeded { node_announcement });
+		}
```
Originally, I had this in `process_events`, but I moved it to `timer_tick_occurred` so the timeout could be in terms of ticks. However, it has occurred to me that having it here means we may wait up to 10 seconds before even generating the events, which would result in bad UX. We could instead simply generate events directly in `process_pending_events` and `process_pending_events_async`, dropping the intermediary `pending_events` `Vec`. But, as before, we'd have to define the timeout in something other than ticks, or possibly add another method to `OnionMessageHandler` for calling in `timer_tick_occurred`.

jkczyz replied:
Updated the `OnionMessageHandler` trait to have a `timer_tick_occurred` method so that dropping buffered messages is independent of providing events.

jkczyz commented on lightning/src/events/mod.rs:
```diff
@@ -513,6 +513,23 @@ pub enum Event {
+	/// [`MessageRouter`]: crate::onion_message::MessageRouter
+	/// [`Destination`]: crate::onion_message::Destination
+	/// [`OnionMessageHandler`]: crate::ln::msgs::OnionMessageHandler
+	ConnectionNeeded {
```
It would require adding another parameterization for users to implement (i.e., either a new trait on `OnionMessenger` or one on `DefaultMessageRotuer`), which would then need to bubble up to the `Simple*` aliases.

And since it needs to be async, the trait implementation wouldn't be enough. Users would need to write code to process connection requests in the background from what their trait enqueued. If we wrote some of this logic for them, then we'd need to do it for each language's IO layer. But that wouldn't fit well into any peer management logic that they may have (e.g., LDK Node's `PeerStore`).

Seems much simpler just to surface an event for them so that no additional parameterization and processing code is needed, other than handling one more event variant.

valentinewallace replied:
Could we pipe the needs-connect nodes through `OffersMessageHandler` / `CustomOMHandler` with a new trait method? Then the `ChannelManager` could generate `ConnectionNeeded` events and we wouldn't have to trouble the background processor. I do see Jeff's argument that the current design aligns with the peer manager doing peer stuff though.

jkczyz replied:
Hmm... that wouldn't work for uses of `send_onion_message` or anyone using something other than `ChannelManager` to implement one of those traits.

valentinewallace replied:
The custom trait implementations would need to support connecting to peers outbound, would it not work in some other way? I'm not convinced it's cleaner than the current approach though :/ just nicer in terms of the `PeerMan`/`BgProcessor`.

jkczyz replied:
Not sure that I follow. All custom implementations should work with the current approach of using `OnionMessenger`. Or are you saying a custom implementation of `OnionMessageHandler` (i.e., something other than `OnionMessenger`)? Note that `CustomOnionMessageHandler` and `OffersMessageHandler` are specific to `OnionMessenger`. Any implementations of those work fine with the current approach.

jkczyz replied:
Discussed offline. Updated to have events processed directly from any `OnionMessageHandler` by requiring implementations to also implement `EventsProvider`. I was able to do so without adding another type parameter to `BackgroundProcessor` since the handler can be fetched from `PeerManager`.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -654,24 +710,26 @@ where
 					},
 					hash_map::Entry::Occupied(mut e) => {
-						e.get_mut().push_back(onion_message);
+						e.get_mut().enqueue_message(onion_message);
```
Yeah, I wasn't sure if we should just buffer if we are already expecting a connection. The alternative would be to fail forwarding (silently) even though we'd later have a connection and could complete the forward.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -569,34 +645,61 @@ where
-			return;
+		match self.send_onion_message(path, contents, reply_path) {
+			Err(SendError::InvalidFirstHop(first_node_id, onion_message)) => {
+				let network_graph = self.network_graph.deref().read_only();
```
Ok, let me go ahead and move `NetworkGraph` to `DefaultMessageRouter`. I guess the idea would be to add a `Vec<SocketAddress>` field to `OnionMessagePath`? I used `NodeAnnouncement` since it would give the user more information on deciding whether they want to connect to the node. But perhaps `SocketAddress`es are sufficient. And I suppose

jkczyz replied:
Ok, let me go ahead and move `NetworkGraph` to `DefaultMessageRouter`. I guess the idea would be to add a `Vec<SocketAddress>` field to `OnionMessagePath`? I used `NodeAnnouncement` since it would give the user more information on deciding whether they want to connect to the node. But perhaps `SocketAddress`es are sufficient. And I suppose a `MessageRouter` may want to connect to a private node, so we should support that.

> (and, awkwardly, passing some event through the peermanager to the user to trigger the connection, but really that is a router-originated event)

An alternative could be to parameterize `BackgroundProcessor` with `OnionMessenger` or `MessageRotuer`, but I figured we'd want to avoid adding more parameterization there. Seemed to make sense to have `PeerManager` be responsible as a more higher level abstraction for peer-related needs.

jkczyz replied:
Related, we'll need to handle `sciddir_or_pubkey` somewhere. That is, going from `Destination` containing a `BlindedPath` with and scid to `OnionMessagePath` having a `Destination` containing a `BlindedPath` with a pubkey.

I suppose doing this in a `MessageRouter` is fine. But when wanting to _construct_ such a blinded path, we'll need the `NetworkGraph`, too. This responsibility then may want to also live in `MessageRouter`, but the current draft is to have it in `Router` partly because `ChannelManager` already has one. So this may need to be rethought.

jkczyz commented:
Updated per offline discussion. Let me know if this new approach looks good, and I'll write up some tests.

jkczyz commented:
Latest push is to fix CI. I'll need to rebase to fix the fuzz test.

valentinewallace commented on lightning-background-processor/src/lib.rs:
```diff
@@ -673,9 +686,32 @@ where
+	use lightning::events::EventsProvider;
+
+	let events = core::cell::RefCell::new(Vec::new());
+	peer_manager.onion_message_handler().process_pending_events(&|e| events.borrow_mut().push(e));
```
Maybe after async functions in traits are stabilized, we can add an async event processing method to `EventsProvider` with a default implementation 🤔 I'm just wondering if this could be a temporary workaround.

jkczyz replied:
Yeah, possibly. `EventsProvider` is kinda weird in that up until now it never really needed to be a trait.

valentinewallace reviewed (commented):
Looks good I think. Will give it another glance later.

valentinewallace commented on lightning-background-processor/src/lib.rs:
```diff
@@ -334,6 +340,11 @@ macro_rules! define_run_body {
 				last_freshness_call = $get_timer(FRESHNESS_TIMER);
 			}
+			if $timer_elapsed(&mut last_onion_message_handler_call, ONION_MESSAGE_HANDLER_TIMER) {
+				log_trace!($logger, "Calling OnionMessageHandler's timer_tick_occurred");
```
Looks like it'd be easy to add some minor test coverage for this.

valentinewallace commented on lightning/src/ln/msgs.rs:
```diff
@@ -1580,6 +1580,9 @@ pub trait OnionMessageHandler {
 	/// drop and refuse to forward onion messages to this peer.
 	fn peer_disconnected(&self, their_node_id: &PublicKey);
 
+	/// Performs actions that should happen on startup and roughly every ten seconds thereafter.
```
FWIW, docs tend to be more extensive on this method elsewhere. You may have been waiting for a concept ACK first though.

valentinewallace reviewed (commented):
I'd be good with a rebase to fix CI

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -509,25 +636,54 @@ where
+
+		match message_buffers.entry(first_node_id) {
+			hash_map::Entry::Vacant(e) => match addresses {
+				None => Err(SendError::InvalidFirstHop(first_node_id)),
```
`first_node_id` isn't necessarily a peer anymore. `DefaultMessageRouter` will look up the node in the `NetworkGraph` if it isn't a peer. More generally, a `MessageRouter` can return a partial path back from the `Destination`, and we'll happily generated a `ConnectionNeeded` event with the first node in such a path.

jkczyz replied:
@TheBlueMatt Any thoughts on this?

jkczyz replied:
The alternative to the graph lookup is storing them on connection from the `Init` message. Either way, we can easily do it later given we don't persist any of this data.

jkczyz replied:
Oh, right. I incorrectly remembered the purpose of those addresses.

jkczyz reviewed (commented):
Added tests and fixed a bug that was preventing the timer from ticking once the `ConnectionNeeded` event was generated. 🤦‍♂️

valentinewallace commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -509,25 +636,54 @@ where
+
+		match message_buffers.entry(first_node_id) {
+			hash_map::Entry::Vacant(e) => match addresses {
+				None => Err(SendError::InvalidFirstHop(first_node_id)),
```
The docs say "The first hop does not support onion message forwarding." But we don't really know that they don't advertise OM support at this point, we only know that they're not connected to us.

jkczyz reviewed (commented):
Will address the nits while waiting on responses for some outstanding comments.

jkczyz commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -427,51 +462,97 @@ fn many_hops() {
+	// Buffer an onion message for a disconnected peer
+	disconnect_peers(&nodes[0], &nodes[1]);
+	assert!(nodes[0].messenger.next_onion_message_for_peer(nodes[1].node_id).is_none());
+	nodes[0].messenger.send_onion_message(message, destination, None).unwrap();
```
This would causes us to hit the debug assertion in `dequeue_message`. We don't prevent someone from calling `next_onion_message_for_peer` for a disconnected peer. Perhaps we should return `None`? Calling it means the `PeerManager` state is out-of-sync with the `OnionMessenger` state.

jkczyz commented:
> Needs rebase, but would be happy to see it squashed.

Rebased and squashed. Had to refactor the sending methods slightly to log onion message contents.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -568,15 +581,60 @@ where
+			Ok(SendSuccess::Buffered) => {
+				log_trace!(self.logger, "Buffered onion message {}", log_suffix);
+			},
+		}
```
Ugh, this needs to be refactored a little more. Otherwise, not all the errors will be logged.

jkczyz commented:
Latest push should fix the `cargo doc` failure in CI.
